### PR TITLE
Add an ADAM-formatted file for testing

### PR DESCRIPTION
**Pull request by laserson**
File generation is included in the commit message.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java [hunk 1]
```diff
@@ -164,8 +165,18 @@ public SAMFileHeader getHeaderForReads() {
      * @return all reads from our reads input(s) as a {@link JavaRDD}, bounded by intervals if specified.
      */
     public JavaRDD<GATKRead> getReads() {
-        // If no intervals were specified (intervals == null), this will return all reads (mapped and unmapped)
-        return readsSource.getParallelReads(readInput, intervals);
+        // TODO: This if statement is a temporary hack until #959 gets resolved.
+        if (readInput.endsWith(".adam")) {
+            try {
+                return readsSource.getADAMReads(readInput, getHeaderForReads());
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/GATKSparkTool.java]: can we get reads from a set of intervals?

**Comment by akiezun** [comment]
lgtm other than the question of intervals.

**Comment by laserson** [comment]
@akiezun Added support for intervals.  But all this stuff should be refactored.  An inefficiency in the ADAM case here is that a `SAMRecord` object must be constructed for each `AlignmentRecord` that is read in before we decide whether it is even in a target `Interval` or not.

**Comment by akiezun** [comment]
thanks Uri. Is that a limitation of the format or of the current code on our side? That is, does the ADAM format support interval queries without having to read the whole file?

**Comment by laserson** [comment]
@akiezun It has to do with the code, one part laziness in avoiding to implement a bona fide `GATKRead` adapter for `AlignmentRecord`, and one part due to the interval filter operating only on `SAMRecord`s (`samRecordOverlaps`).  Assuming ADAM does become a more important use case to support, we can refactor to improve this.

**Comment by laserson** [comment]
(Also note that the format is not relevant in this particular case, because the filtering step is occurring in an RDD transformation on a record-by-record basis.)

**Comment by akiezun** [comment]
hmm, but does an ADAM file have an index or something that you can ask for reads from a given interval? Do we need to read in all the reads and filter them one by one?

**Comment by laserson** [comment]
I terms of comparing implementations in this particular instance, the entire BAM file is also being scanned, so there is no additional limitation in using Parquet.

In general, though, there are some additional complications for typical Hadoop data sets due to their distributed nature.  There are multiple methods for accelerating lookups, depending on the particular application and what kind of latency you need.  A few of them:
* Data set partitioning in the style of Hive, where you split your data set into a directory hierarchy that's based on the values of one of the columns.  This is like building an index, and should be done on cols that feature in lot of predicates.  In the quince repo that we're using for ingest, we are partitioning the data based on genome locus.  This makes it easy to access the data only from the locus of interest.
* Parquet supports predicate pushdown and column stats on its row chunks, so the more you homogenize the data (e.g., by sorting), the more likely it is you can skip large blocks of data.
* The parquet file format supports the concept of indexing (though I don't think it is implemented yet in any of the packages that read/write it)
* For particular applications, you can also use a different storage backend like HBase or Kudu that allow very rapid point/range queries at scale.  I believe GEL is planning on trying out HBase for some of their applications
* The Hammerbacher lab and the ADAM folks are also working on tools for BAM file indexing and visualization that scales on Hadoop.  I think the projects are cycledash/pileup.js and mango, respectively.

**Comment by akiezun** [comment]
thanks @laserson. That sounds complicated. How do people work with those files then - only using them for whole genome traversals? That sounds very limiting.

The code looks fine. Please rebase squash and merge. :+1:

**Comment by laserson** [comment]
I think is sounds more complicated than it is.  From the user's perspective, there is a "data file" which is really a directory in HDFS, and they just use their tool of choice as usual.  It should know how to deal with the "indexing" as appropriate.

**Pull request merged**